src/components/PluginLoader.jsx

```diff
@@ -64,35 +105,72 @@ export const PluginLoader = ({
         []
     )
 
+    // todo: add query string to entrypoint (e.g. maps /dhis-web-maps/?currentAnalyticalObject=true)
+    // Can use `urlObject.searchParams.append('redirect', 'false')`
+    // (or is this a backend thing?)
     React.useEffect(() => {
         if (!appsInfoQuery.data) {
             return
         }
+        // Performed async to test for index/app entrypoint
         const asyncWork = async () => {
             // for testing: params.appName === 'localApp' ? 'http://localhost:3001/app.html'
-            const newPluginSource = await newGetPluginSource(
+            const newPluginEntrypoint = await getPluginEntrypoint(
                 params.appName,
                 appsInfoQuery.data.appMenu.modules,
                 baseUrl
             )
-            setPluginSource(newPluginSource)
+            setPluginEntrypoint(newPluginEntrypoint)
         }
         asyncWork()
     }, [params.appName, baseUrl, appsInfoQuery.data])
 
-    if (!pluginSource) {
+    const pluginHref = React.useMemo(
+        () =>
+            // An absolute URL helps compare to the location inside the iframe:
+            new URL(
+                pluginEntrypoint + '?redirect=false' + location.hash,
+                window.location
+            ).href,
+        [pluginEntrypoint, location.hash]
+    )
+
+    const handleLoad = React.useCallback(
+        (event) => {
+            // If we can't access the new page's Document, this is a cross-domain page.
+            // Disallow that; return to previous plugin state
+            // todo: figure out some better UI
+            if (!event.target.contentDocument) {
+                setRerenderKey((k) => k + 1)
+                // Timeout so the plugin can reset before this shows
+                setTimeout(() => {
+                    alert(
+                        'Failed to navigate to a URL outside DHIS2. Returned to previous page.'
+                    )
+                }, 400)
+                return
+            }
+
+            if (handleExternalNavigation(event, pluginHref)) {
+                return
+            }
+            injectHeaderbarHidingStyles(event)
+            watchForHashRouteChanges(event)
+        },
+        [pluginHref]
+    )
+
+    if (!pluginHref) {
         return 'Loading...' // todo
     }
 
     return (
         <Plugin
-            width={'100%'}
-            height={'100%'}
-            // todo: only for apps without header bars
-            // height={'calc(100% - 48px)'}
+            className={styles.flexGrow}
```

Comment:
Second part of the fix to have plugin size be (page height - header bar height), instead of just page height (which made it too tall)